hack/vf-netns-switcher.sh

```diff
@@ -0,0 +1,369 @@
+        if [[ "$phys_port_name_pf_index" != "${pf_port_names[$pf_name]:1}"  ]]; then
+            continue
+        fi
+        echo "Switching VF representor $interface of PF $pf_name to netns $worker_netns"
+        switch_vf $interface $worker_netns
+    done
+}
+
+switch_interface_vfs(){
+    local pf_name="$1"
+    local worker_netns="${2:-$netns}"
+    local pci="$3"
+
+    vfs_list=$(ls /sys/bus/pci/devices/$pci | grep virtfn)
+
+    if [[ -z "${vfs_list}" ]];then
+        echo "Warning: No VFs found for interface $pf_name!!"
+        return 0
+    fi
+
+    for vf in $vfs_list;do
+        local vf_interface="$(ls /sys/bus/pci/devices/$pci/$vf/net)"
+
+        if [[ -n "$vf_interface" ]];then
+            echo "Switching $vf_interface to namespace $worker_netns..."
+            sleep 2
+            if ! switch_vf "$vf_interface" "$worker_netns";then
+                echo "Error: could not switch $vf_interface to namespace $worker_netns!"
+            else
+                echo "Successfully switched $vf_interface to namespace $worker_netns"
+            fi
+        fi
+    done
+}
+
+read_confs(){
+    local conf_file="$1"
+
+    let number_of_netns=$(jq length "${conf_file}")-1
+
+    for index in $(seq 0 $number_of_netns);do
+        netnses+=("$(jq -r .[${index}].netns $conf_file)")
+        let number_of_pfs=$(jq .[$index].pfs $conf_file | jq length)-1
+        for pf_index in $(seq 0 $number_of_pfs);do
+            pfs[${netnses[-1]}]+="$(jq -r .[$index].pfs[$pf_index] $conf_file) "
+        done
+    done
+}
+
+variables_check(){
+    local status=0
+
+    check_empty_var "netns"
+    let status=$status+$?
+    check_empty_var "pfs"
+    let status=$status+$?
+
+    return $status
+}
+
+check_empty_var(){
+    local var_name="$1"
+
+    if [[ -z "${!var_name[@]}" ]];then
+        echo "Error: $var_name is empty..."
+        return 1
+    fi
+
+    return 0
+}
+
+main(){
+    while true;do
+        for netns in ${netnses[@]};do
+            switch_pfs "$netns" "${pfs[$netns]}"
+            sleep 2
+            switch_netns_vfs "$netns"
+            sleep 2
+            switch_netns_vf_representors "$netns" "${pfs[$netns]}"
+        done
+        sleep $TIMEOUT
+    done
+}
+
+if [[ -n "$conf_file" ]];then
+    unset netnses
+    unset pfs
+
+    declare -a netnses
+    declare -A pfs
+
+    read_confs "$conf_file"
+fi
+
+variables_check
+let status=$?
+if [[ "$status" != "0" ]];then
+    echo "Error: empty var..."
+    exit $status
+fi
+
+for netns in ${netnses[@]};do
+    netns_create "$netns"
+    let status=$status+$?
+    if [[ "$status" != "0" ]];then
+        echo "Error: failed to create netns..."
+        exit $status
+    fi
+done
+
+for netns in ${netnses[@]};do
+    get_pcis_from_pfs "$netns" "${pfs[$netns]}"
+    get_pf_switch_dev_info "$netns" "${pfs[$netns]}"
+done
+
+if [[ "${#pcis[@]}" == "0" ]];then
+    echo "Error: could not get pci address of interface $pf!!"
+    exit 1
+fi
+
```

Comment:
is it desirable to have a trap so that when the service exits it will move netdevs back to default namespace ?

Reply:
how about this:

1. create namespace with random name
2. move VFs from ns in config to this namespace
3. delete namespace from 1(will put netdev back in root namespace)

Reply:
Done, can you have a look